File: sorting/cycleSort/missingNumber.java
// https://leetcode.com/problems/missing-number/

package sorting.cycleSort;

public class missingNumber{
    public static void main(String[] args) {
        SolutionMN solution = new SolutionMN ();
        System.out.println(solution.missingNumber(new int[]{0,1}));
        System.out.println(solution.missingNumber(new int[] {3,0,1}));
    }
}

class SolutionMN {
    public int missingNumber(int[] nums) {
        int n = nums.length;
        int i = 0;
        while(i<n ){
            if(nums[i]==i || nums[i]==n) i++;
            else swap(nums, i, nums[i]);
        }
        for (int j = 0 ; j< n ; j++) if (nums[j] != j) return j;
        return n;
        
    }
    public void swap (int [] array , int currentIndex, int correctIndex){
        int temp= array[ correctIndex];
        array[correctIndex]= array[currentIndex];
        array[currentIndex]= temp;
        
    }
}

File: sorting/customSort/missingNumber.py
# https://leetcode.com/problems/missing-number/
class Solution:
    
    # TC :  O(N)
    # SC : O(1)
    def missingNumber(self, nums) -> int:
        
        
        for i in range(len(nums)):
            while nums[i]<len(nums) and nums[i] != i:
                numberAtIndexI = nums[i]
                nums[i], nums[numberAtIndexI]= nums[numberAtIndexI], nums[i]
                
        for i in range(len(nums)):
            if nums[i]!=i:
                return i
    
        return len(nums)
    
    # TC : O(N)
    # SC : O(1)
    def missingNumber(self, nums) :
        
        summation = 0
        maximum = 0
        is_zero = False
        
        for i in nums:
            summation+=i
            maximum = max(i,maximum)
            if is_zero==False: is_zero = (i==0)
        
        missingNumber = maximum*(maximum+1)//2 -summation
        
        return maximum+1 if missingNumber==0 and is_zero else missingNumber 
    
    
